### DISCUSSION TITLE
Discussion comparing coral and iGPU (OpenVino) as detector, pros & cons

### DISCUSSION THREAD

**Discussion by quenthal** [blakeblackshear/frigate#5742] [General]
I've frigate running on minipc with Iris XE iGPU. I 've also Coral USB and Coral mPCI available.

I've been comparing these within Proxmox LXC, where Frigate container is running. 7 cameras.

I get interference slight below 8ms with mPCI, USB version nearly the same. With OpenVino with iGPU I get interference around 10-12ms. 

With OpenVino CPU on LXC with 4 virtual cores CPU usage (looking from Proxmox) is generally around 12-13% with recurring spikes up to 40%.

With Coral the cpu is generally also around 12-13%, but the spikes are slightly less recurring, up to 32%, sometimes up to 40%.

Other points to compare are of course the quality of the detecting between OpenVino ja Coral. I haven't yet got any definitive data on  this.

One big, major I would say thing to compare is also power consumption. I haven't yet unplugged for these comparisons mPCI Coral physically, but it would be also intresting to see if Coral consumes more power than what the increased usage of iGPU and maybe CPU causes.

What are your findings or thoughts on these matters?

**Comment by NickM-27** [comment]
Coral is more power efficient. Their accuracy also has nothing to do with the device itself as that is only dependent on the model

**Reply by quenthal** [reply]
Thanks!

**Reply by bazylhorsey** [reply]
I think that is a somewhat technically correct assessment @quenthal. However, there are a lot of limitations on models that can be ran on coral due to hardware limitations. In computer vision specifically, it's really hard to get results you see that are crazy accurate like yolo v8.
I'd be curious to see some kind of accuracy/latency graph for edge Intel/coral accessible models. I'm trying to create custom models that are both very accurate, and very fast. This proves difficult in an edge ecosystem, at the same time, these models are impossible to run anywhere else in the same cost point.

**Comment by gszigethy** [comment]
I have tested a few setups and the difference is actually not that big between Coral and OpenVino which is a big surprise for me too.

I have 9 cameras monitored, the detect streams are 1280x720, the record streams are various, but between FHD and 4k, all of them using h.264. Storage backend in each case is a CIFS mounted share on the same switch with an SSD backend, all ports are 1Gb. Cpufreq governor set to ondemand, powertop auto-tune.

- Raspberry Pi 4, USB Coral as detector: 9W (including ~1w of a POE splitter and a tiny fan in the case): CPU 40-50%, inference speed 20-30 ms
- NUC8i5BEK, Proxmox 7, Frigate in a docker container on a Debian LXC, GPU passthrough, OpenVino: 25-35W, CPU 20-30%, GPU 10%, inference 7-9 ms
- NUC11TNHv50L, Proxmox 7, Frigate in a docker container on a Debian LXC, GPU passthrough, OpenVino: 16-26W, CPU 20-30%, GPU 10%, inference 7 ms
- NUC11TNHv50L, Proxmox 7, Frigate in a docker container on a Debian LXC, GPU passthrough, USB Coral: 20-32W, CPU 7-15%, GPU 2%, inference 7-8 ms

Some observations:
- Pi is the lowest power by far and can handle the above number of cameras, but tested only with Frigate 11 (fr 12 seem to had various issues on pi, didn't have the time to debug yet).
- Pi, while low power and can handle the NVR capabilities, watching events and using the interface in general is not as snappy as a NUC, obviously.
- Accuracy of OV and Coral detection are very similar. I've compared a day of events on the most busy camera (facing a street) and it was almost identical. They mostly miss events nighttime when the camera picture is worse, lower fps, the screen is blurry, etc. But the records are still there as the movement is detected, just not detected as an event (so you can still capture it, but less convenient to check what happened).
- Confidence of OV is much higher, ie. OV events are in the high 90%, Coral is more distributed in the range of 60-100%.
- The power consumption is stable on the pi, 9w, probably because it goes on max fequency already at 50% cpu usage. The power draw of the NUCs can double based on amount of movement and event detection. For instance, on a windy day can go significantly higher.
- Using poor quality cable for the Coral and connecting it on USB2 (480M in lsusb) makes the inference speed and the CPU usage go kaboom.

In general I'm very much pleased with the OpenVino on a NUC, especially if there are Coral TPU shortages, it's a perfectly viable route, also considering that ffmpeg on amd64 is much less picky then on a pi. It eats more power, but you get a lot more too.

**Reply by NickM-27** [reply]
Just wanted to note

> - Accuracy of OV and Coral detection are very similar. I've compared a day of events on the most busy camera (facing a street) and it was almost identical.
> - Confidence of OV is much higher, ie. OV events are in the high 90%, Coral is more distributed in the range of 60-100%.

These are factors of the model being run and have nothing to do with the hardware that is used for detection (coral, OV, etc.)

**Reply by mac587** [reply]
Am I reading this correctly? That coral is basically redundant on anything with intel iGPU from 6th gen onwards?

**Reply by cdennis12280** [reply]
Think so I'm running on old dell laptop i5 6tg gen. 4 cams, inf speed around 12-15ms.

Now need to source a cheap tiny pc with decent cpu and proxmox my setup.

**Reply by charettepa** [reply]
> Am I reading this correctly? That coral is basically redundant on anything with intel iGPU from 6th gen onwards?

I am wondering the same thing, have there been any discoveries/answers to this?

**Reply by NickM-27** [reply]
it depends what your priorities are, coral will still be lower power usage and for some people electricity is expensive enough that is the priority.

**Reply by charettepa** [reply]
yes my concern is power usage and keeping the cpu load low

was just reading another thread where people have n100 and n200
they are reporting lower power usage with igpu than with coral
with comparable performance

im just not sure anymore
i have an i3-n305 (8 core gen 13 cpu)
the coral is currently 134.99 
i think i would need to see major performance difference and much lower power consumption
in order to pay for itsself
even worse if it uses more power


I only have 3 cameras and really only one of them will see much motion
will this play a role

**Reply by NickM-27** [reply]
why don't you just try using openvino? doesn't hurt anything to try and see how it performs

**Reply by charettepa** [reply]
oh i plan i trying very shortly. 

i have never used frigate before
just trying to gather info.
is there a breaking point where the coral makes more sense
like, i dont know.... more than X cameras 
and condition X

also if im using open vino for frigate, can i still use the gpu for something else, or no?
as i will also be using the gpu for plex/jellyfin
if the gpu can not be shared that might be an issue

I am coming from a Synology NAS and now running Unraid
i never had object detection and could go without it if need be
frigate just seemed like the package to use moving forward

**Reply by NickM-27** [reply]
yes, GPU can be shared as long as it is across docker containers. GPU can not be shared with multiple VMs at the same time

**Reply by xavan83** [reply]
> yes my concern is power usage and keeping the cpu load low
> 
> was just reading another thread where people have n100 and n200 they are reporting lower power usage with igpu than with coral with comparable performance
> 
> im just not sure anymore i have an i3-n305 (8 core gen 13 cpu) the coral is currently 134.99 i think i would need to see major performance difference and much lower power consumption in order to pay for itsself even worse if it uses more power
> 
> I only have 3 cameras and really only one of them will see much motion will this play a role

I have one to sell if you want :-D

**Reply by charettepa** [reply]
OK. I ended up buying one. So far i think it was worth it. My CPU never goes above 7% usually sits at 4% even with large model enabled. Even then the Coral TPU CPU is rarely over 5%. seems verry efficient. I will measure power usage and report back.

**Reply by NickM-27** [reply]
just to be clear, the coral tpu does not have a %, it simply runs single inferences as fast as possible.

**Reply by patrykk** [reply]
> yes, GPU can be shared as long as it is across docker containers. GPU can not be shared with multiple VMs at the same time

Intel iGPU can be shared using SR-IOV or GVT-g but it have to be configured on host machine manually. I have had successes with both of them. Nvidia block GPUs and it is more difficult to do but not possible. I don't know status of the AMD gpus.

**Comment by comfia** [comment]
Greetings, friends. I have a Beelink EQ13 with an Intel 12th Gen Alder Lake-N100, 16GB RAM, and 500GB storage. I want to install Frigate using Docker on Ubuntu, with 5 cameras. Will it work with OpenVINO, or do I absolutely need to buy a Coral TPU? Thank you.

**Reply by hawkeye217** [reply]
OpenVINO should work just fine. No need for a Coral.

**Reply by comfia** [reply]
Thank you so much for your kind and helpful response. I really appreciate your time and support!

**Reply by charettepa** [reply]
i have an n305 and with only 2 cameras mine was using too much cpu always at 20+% and hitting 40's
cpu usage was reduced with the use of the coral never going above 7 and idling at 4
